tethys_apps/base/app_base.py

```diff
@@ -131,6 +136,82 @@ def url_patterns(self):
 
         return self._url_patterns
 
+    @property
+    def handler_patterns(self):
+        """
+        Generate the url pattern lists for  app and namespace them accordingly.
+        """
+        if self._handler_patterns is None:
+            is_extension = isinstance(self, TethysExtensionBase)
+
+            handler_patterns = {'http': dict(), 'websocket': dict()}
+
+            if hasattr(self, 'url_maps'):
+                url_maps = self.url_maps()
+
+            for url_map in url_maps:
+                if url_map.handler:
+                    namespace = self.namespace
+
+                    if namespace not in handler_patterns['http']:
+                        handler_patterns['http'][namespace] = []
+
+                    if namespace not in handler_patterns['websocket']:
+                        handler_patterns['websocket'][namespace] = []
+
+                    # Create django url object
+                    if isinstance(url_map.handler, str):
+                        root_handler_path = 'tethysext' if is_extension else 'tethysapp'
+                        full_handler_path = '.'.join([root_handler_path, url_map.handler])
+                        handler_parts = full_handler_path.split('.')
+                        module_name = '.'.join(handler_parts[:-1])
+                        function_name = handler_parts[-1]
+                        try:
+                            module = __import__(module_name, fromlist=[function_name])
+                        except Exception as e:
+                            error_msg = 'The following error occurred while trying to import the handler function ' \
+                                        '"{0}":\n {1}'.format(url_map.handler, traceback.format_exc(2))
+                            tethys_log.error(error_msg)
+                            raise e
+                        try:
+                            handler_function = getattr(module, function_name)
+                        except AttributeError as e:
+                            error_msg = 'The following error occurred while trying to access the handler function ' \
+                                        '"{0}":\n {1}'.format(url_map.handler, traceback.format_exc(2))
+                            tethys_log.error(error_msg)
+                            raise e
```

Comment:
Lines 163-184 are virtually identical to lines 109-130. Please generalize into a  private method in this class. Something along the lines of `_resolve_function_ref(ref, ref_type)` used like `self._resolve_function_ref(url_map.controller, 'controller')`. The `ref_type` parameter would be used in the `error_msg`s.